extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -0,0 +1,363 @@
+import {CSS} from '../../../build/amp-mega-menu-0.1.css';
+import {Keys} from '../../../src/utils/key-codes';
+import {Layout} from '../../../src/layout';
+import {Services} from '../../../src/services';
+import {
+  closestAncestorElementBySelector,
+  isRTL,
+  scopedQuerySelector,
+  scopedQuerySelectorAll,
+  tryFocus,
+} from '../../../src/dom';
+import {dev, userAssert} from '../../../src/log';
+import {isExperimentOn} from '../../../src/experiments';
+import {setModalAsClosed, setModalAsOpen} from '../../../src/modal';
+import {toArray} from '../../../src/types';
+
+/** @const {string} */
+const TAG = 'amp-mega-menu';
+
+/** @const {string} */
+const SCREEN_READER_CLOSE_LABEL = 'Close the mega menu';
+
+/**
+ * A mega menu component suitable for displaying large collections of links on
+ * desktop/tablets. An amp-list can be nested inside this component to enable
+ * template rendering via dynamically fetched data.
+ */
+export class AmpMegaMenu extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Array<!Element>} */
+    this.items_ = null;
+
+    /** @private {number} */
+    this.itemCount_ = 0;
+
+    /** @private {?Element} */
+    this.expandedItem_ = null;
+
+    /** @private {?Element} */
+    this.maskElement_ = null;
+
+    /** @private @const {!Document} */
+    this.document_ = this.win.document;
+
+    /** @private @const {!Element} */
+    this.documentElement_ = this.document_.documentElement;
+
+    /** @private {?../../../src/service/action-impl.ActionService} */
+    this.action_ = null;
+
+    /** @private {number|string} */
+    this.prefix_ = element.id || Math.floor(Math.random() * 100);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout === Layout.FIXED_HEIGHT;
+  }
+
+  /** @override */
+  buildCallback() {
+    // TODO(#24898): remove this assert when cleaning up experiment post launch.
+    userAssert(
+      isExperimentOn(this.win, 'amp-mega-menu'),
+      'Turning on the amp-mega-menu experiment is necessary to use the amp-mega-menu component.'
+    );
+
+    this.action_ = Services.actionServiceForDoc(this.element);
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.registerMenuItems_();
+    // items may not be present after build if dynamically rendered via amp-list,
+    // in which case register them after DOM update instead.
+    this.element.addEventListener(AmpEvents.DOM_UPDATE, () => {
+      this.registerMenuItems_();
+    });
+
+    const mask = this.document_.createElement('div');
+    mask.classList.add('i-amphtml-mega-menu-mask');
+    mask.setAttribute('aria-hidden', 'true');
+    // append mask to header so that all children of header appear above the mask
+    const maskParent =
+      closestAncestorElementBySelector(this.element, 'header') || this.element;
+    maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
+    maskParent.appendChild(mask);
```

Comment:
How does this interact with other masks?

Reply:
The mask will be below masks for components like sidebar or lightbox gallery. Are there specific concerns you have with the interaction?

Reply:
Here are a couple of interesting use cases to test (or ban via validator): 
1. `amp-lightbox-gallery` triggered by a lightbox-ed image within `amp-mega-menu`. 
2. `amp-sidebar` opened by a click action while `amp-mega-menu` is open. 
3. `amp-lightbox` triggered by clicking on something inside `amp-mega-menu`. I can actually imagine people wanting to do this for popup offers, though I'm unclear on whether we want to encourage this kind of UX. 
4. Multiple `amp-mega-menu`s done for purpose of building multiline menus. 
5. `amp-mega-menu` inside `amp-lightbox`. If this is not a valid use case, we should ban it via validator. 

We should (feel free to do so in a follow up PR): 
1. Verify these cases don't break or result in awkward / bad-looking UX. 
2. If they do, either fix or restrict them.

Reply:
Good call! I'll go through these cases and include relevant manual tests as part of validation pr.